### ISSUE TITLE
Add support for the Wayland display server

### ISSUE THREAD

**Issue by akien-mga** [godotengine/godot-proposals#990]
**Describe the project you are working on:**
Godot's platform ports.

**Describe the problem or limitation you are having in your project:**
None directly as X11 still works decently on recent Linux distros, but we need to add support for the more modern Wayland display server eventually.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
Wayland is a newer, more modern display server that can be used on Linux and likely some BSD-derived OSes too (now or in the future).

Many Linux distros / DEs now run on Wayland by default, and Godot only works thanks to the `xwayland` compatibility layer.

Having builtin Wayland support will give us access to a much cleaner API as well as more advanced features for window management, etc.

Supersedes https://github.com/godotengine/godot/issues/576 - see discussion there for relevant technical details.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
I'm not an expert on Wayland so I'll let others fill in and explain what are the gains we can expect from having built-in Wayland support.

From a technical implementation point of view, the following points are important:
- Implement Wayland support was intentionally kept on hold until we would split the OS and DisplayServer components, so that we can have a Linux/*BSD platform port with more than one DisplayServer implementations. This has now been done in the `master` branch, and thus Wayland support can be implemented for Godot 4.0. There's no intention to backport it to earlier releases.
- There are many issues on all platforms following the DisplayServer split and implementation of multi-window support (https://github.com/godotengine/godot/issues/37734). It is recommended to wait for most of those issues to be fixed before starting implementing a new DisplayServer (though help is very welcome from Wayland aficionados to first fix the outstanding X11 issues after this refactoring).
- Wayland and X11 support should be implemented in a way that works with a single Godot binary, i.e. the relevant libraries should probably be `dlopen`-ed so that the same binary can be used on systems with only X11, only Wayland, or both installed.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
No.

**Is there a reason why this should be core and not an add-on in the asset library?:**
This can only be done in core.

**Comment by akien-mga** [comment]
Relevant PRs with initial Wayland implementation but that pre-dated the OS/DisplayServer split and thus couldn't be merged:

https://github.com/godotengine/godot/pull/23426 (@toger5)
https://github.com/godotengine/godot/pull/27463 (@ddevault)

**Comment by ssokolow** [comment]
> Having builtin Wayland support will give us access to a much cleaner API as well as more advanced features for window management, etc.

Most notably for me, denying Wayland applications un-privileged access to the "confuse the window manager about whether this resolution change is permanent enough that windows should be repositioned to keep them within the bounds of the desktop" API.

**Comment by Calinou** [comment]
@ssokolow Godot never attempts to change the screen resolution, so this shouldn't be an issue in the current state of things. Still, Wayland has many other benefits.

**Comment by ssokolow** [comment]
Oh, certainly. It's more that so many *other* games screw that up, or misbehave when they discover that I've removed standard resolutions like `800x600` from the list of valid fullscreen resolutions that it's by far the most noteworthy Wayland feature for me on a machine which is going to be playing games.

(eg. The game Dungeons of Dredmor will refuse to accept any windowed resolution that's not a valid fullscreen resolution. I've been meaning to write an SDL `LD_PRELOAD` hook for ages to lie about the list of valid fullscreen resolutions and then force windowed mode if one of the injected ones is chosen.)

**Comment by carwyn** [comment]
Earlier this week I read the updates in the [X11 display server progress report](https://godotengine.org/article/x11-display-server-progress-report) article which does mention Wayland being "still not widely used". 

Meanwhile today in this article: [It's Time To Admit It: The X.Org Server Is Abandonware](https://www.phoronix.com/scan.php?page=news_item&px=XServer-Abandonware) there are various snippets that give an idea of current state of play for Wayland and X11. Notably they mention that Ubuntu 21.04 may well be the version that defaults back to Wayland.

Currently Fedora, RHEL since version 8 and Debian since version 10 I believe all default to Wayland over X11. With Ubuntu possibly joining them next year I'm not sure that "still not widely used" is entirely accurate :)

Posting to provide the second article above (which has links to relevant conversations) as evidence really.

**Comment by Calinou** [comment]
> Currently Fedora, RHEL since version 8 and Debian since version 10 I believe all default to Wayland over X11. With Ubuntu possibly joining them next year I'm not sure that "still not widely used" is entirely accurate :)

Keep in mind that on NVIDIA graphics cards, the installers will often fall back to X11.

**Comment by pouleyKetchoupp** [comment]
Just adding actual data to the conversation. It's great that things are progressing, but the current state seems to be that most gamers are on X11:
https://www.gamingonlinux.com/users/statistics#SessionType-top

So I don't think it's exaggerated to state that Wayland is not widely used at the moment. That said, now that the X11 display server is in a better shape, contributions for Wayland support are welcome :)

**Comment by carwyn** [comment]
> Just adding actual data to the conversation. ...
> https://www.gamingonlinux.com/users/statistics#SessionType-top

The distribution statistics in there are worth looking at too with most being Ubuntu based. If Ubuntu does indeed jump with 21.04 then there would be a fairly big swing fairly quickly potentially.

Either way, more than happy to help test/debug Wayland support :)

**Comment by Calinou** [comment]
> If Ubuntu does indeed jump with 21.04 then there would be a fairly big swing fairly quickly potentially.

Most Ubuntu users stick to LTS versions nowadays.

**Comment by dvergeylen** [comment]
> Just adding actual data to the conversation. It's great that things are progressing, but the current state seems to be that most gamers are on X11:
> https://www.gamingonlinux.com/users/statistics#SessionType-top

Quiet funny, 2 users reported they use Cinnamon as their Wayland Desktop although Cinnamon doesn't work under Wayland (see [here](https://github.com/linuxmint/cinnamon/issues/8512) and [here](https://github.com/linuxmint/Cinnamon/issues/5201)).

I would love to see native Wayland support but I have to admit nVidia [recently submitted patches](https://gitlab.freedesktop.org/xorg/xserver/-/merge_requests/587) to support hardware accelerated rendering under XWayland, with performances "roughly on-par with native X11".

Yet another argument to say that Wayland support is not *that* critical by now 🤷. Probably at least not before https://github.com/godotengine/godot/issues/37734 .

**Comment by dvergeylen** [comment]
> If Ubuntu does indeed jump with 21.04 then there would be a fairly big swing fairly quickly potentially.

[Ubuntu 21.04 will indeed default to Wayland](https://discourse.ubuntu.com/t/trying-wayland-by-default-again/20575), paving the way for 22.04 LTS. Nvidia users will still default to Xorg for now ("but hopefully that situation will be resolved before the LTS") 😃.

**Comment by Sslaxx** [comment]
https://www.gamingonlinux.com/2021/04/xwayland-work-for-hardware-accelerated-nvidia-support-has-been-merged-in - sounds like it'll be ready for Nvidia users on Ubuntu by 21.10 (assuming they don't release a point update for 21.04 with it), so high probability for 22.04.

**Comment by CrezyDud** [comment]
> I’m currently running ubuntu 21.04 and it defaults to Wayland.

how?
is it implemented?
doesn't for me.

WE NEED WAYLAND SUPPORT
XWAYLAND IS A BUGGY MESS
PLEAAEEAEAEAEEAEAEAESE

**Comment by Calinou** [comment]
> how? is it implemented? doesn't for me.

It runs via XWayland – Godot doesn't have native Wayland support yet. Native Wayland support is unlikely to be added to 4.0 due to time constraints, but it may be added in a future 4.x release.

**Comment by iMonZ** [comment]
This is one of the last issues before I can use the godotendine.
Hope this will be ready soon.

**Comment by Calinou** [comment]
@iMonZ Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by Riteo** [comment]
Hi! I'd be glad to implement this. I already got a working prototype.

**Comment by Riteo** [comment]
@dvergeylen it's still extremely simple and rough, although I got basic window reconfiguration and just handling of the pointer's left click. If you want I can open a draft PR though.

**Comment by ryleu** [comment]
> @ dvergeylen it's still extremely simple and rough, although I got basic window reconfiguration and just handling of the pointer's left click. If you want I can open a draft PR though.

A draft PR implies that *you* want to work on it, so if you don't, you'll want to be clear about that in the description of the PR.

**Comment by Riteo** [comment]
@ryleu yeah, that's the point of my comment. I definitely want to work on it. I would've loved to give you all something complete right away but I was worried that someone, not seeing activity here, would think that nobody was working on it and tried implementing it parallel with me. I didn't want to waste anybody's time, mine included. Regarding the draft PR, I think that I'll open it as soon as I get keyboard support, in order to give people the bare minimum to test it and maybe talk about eventual architectural decisions.

**Comment by toger5** [comment]
@Riteo you did have a look at: https://github.com/godotengine/godot/pull/27463 and https://github.com/godotengine/godot/pull/23426? just to be sure? That is where I and ddevault started working on this.

**Comment by Riteo** [comment]
@toger5 thanks for caring to let me know. I know about the existence of these two PRs and perhaps I may take some stuff from them but for now I didn't really feel the need, as it's pretty much a matter of following the spec and converting to godot's data structures when needed. I didn't want to start from them as the `DisplayServer` split, multiwindow support and Vulkan probably changed the architecture quite a bit, and I still needed to learn the principles behind interactions through Wayland anyways.

**Comment by dvergeylen** [comment]
Riteo's draft PR is available here: https://github.com/godotengine/godot/pull/57025

**Comment by KiwwiPity** [comment]
@akien-mga if I'm not wrong Wayland throws problems with drawing tablets, and `xwayland` doesn't solve that.
Myself had an issue recently with my huion tablet, and as `xwayland` did nothing, needed to switch to X11 to be able to work on Krita.
(not sure if I'm missing something; however is good to keep the tablet topic in mind, there are some apps built using Godot like Lorien that use drawing tablets)

**Comment by Riteo** [comment]
@KiwwiPity what problems, exactly?

**Comment by ssokolow** [comment]
I think I remember reading Phoronix posts about how they're still in the process of implementing various parts of tablet support in the compositors.

(Which would be why using XWayland did nothing. The TODOs sit between it and the hardware.)

**Comment by KiwwiPity** [comment]
@Riteo being the drivers installed and all set up correctly, on Wayland sessions when move the pen, doesn't move the cursor, so it doesn't do nothing at practice.

**Comment by Riteo** [comment]
@KiwwiPity I see, I thought it was some protocol issue. Weird though, I thought that Wayland used libinput for input handling and all old drivers should've worked fine.

**Comment by ryleu** [comment]
I personally have had no issues with a tablet stylus (my laptop has built-in support, it’s not a normal touchscreen stylus) on Wayland. Both Krita and Rnote work perfectly with it. This also seems like a relatively small thing for Godot, considering that the vast majority of games don’t need a stylus. For the stylus games, they can just use X.

I think the biggest blocker for Wayland games is the lack of support on older NVIDIA cards. For example: my GTX 780.

**Comment by KiwwiPity** [comment]
@Riteo Huion tablets have not official support for Linux, the drivers I use are community made, so the problem could be there.

**Comment by snoopdouglas** [comment]
Will work on this also implement HiDPI awareness on Linux?

**Comment by HalanoSiblee** [comment]
2023 and still no wayland support it's disgusting and lazy game engine i don't care about it anymore

**Comment by Calinou** [comment]
@HalanoSiblee Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by KiwwiPity** [comment]
@HalanoSiblee Wayland still very problematic in general, even XFCE lacks wayland support until now.
Also those Linux users that own an Nvidia GPU would have problems with Wayland, even if it wasn't from Godot side.
So its a domain problem, more than a Godot problem, properly speaking.
For now it looks like its almost impossible gaming properly or any kind of real-time graphics with Nvidia + Wayland.
If someone knows how fix that just shoot.

**Comment by sonic2kk** [comment]
Supposedly KDE Plasma Wayland works for Nvidia users (with a handful of remaining limitations, see the Nvidia section of the Plasma Wayland showstoppes), but AMD and Intel GPUs work fine for gaming on Wayland. I have been using Plasma Wayland for gaming for well over a year on my 5700XT, and on a couple of Intel laptops, specifically for a mixed-scaling and mixed-refresh-rate setup. 

Some desktop environments, such as KDE, are moving to Wayland as the default. Of course, distributions can still override this like Ubuntu do/did with GNOME. Native Wayland support may gain more interest as a result of this though, mainly for mixed-scale users where KDE's X11 scaling option doesn't help, and XWayland apps are blurry. Other DEs may suffer similar issues as well.

**Comment by sonic2kk** [comment]
Indeed, usually the desktop toolkit takes care of this (GTK, Qt) and/or the underlying tech stack (Chromium for Electron apps) but in Godot's case that won't apply. I believe Wine's Wayland support is also in the same boat as requiring additional work in this regard.

Wayland support is a step in the right direction for scaling support, though, as afaik this kind of mixed-scale stuff can't really be done cleanly on the application side under X11.

**Comment by saolof** [comment]
Possible idea for a slightly more out-there feature that I wish that at least one game engine supported: being able to have Wayland windows float inside a Godot game, essentially using Godot as a window manager.

This is primarily useful for VR productivity applications. VR headsets have the very real potential to be the form factor of the future, combining portability with theoretical maximum screen real estate. The issue has been on the software side, with all headsets so far being completely consumer focused, and my ideal would be to just have a linux distro that uses a game engine acting as the WM.

**Comment by DeedleFake** [comment]
Godot is cross-platform. If it was going to work as a Wayland compositor, I would think that it would have to have similar functionality for Windows, Mac, and others. That seems quite infeasible.

That being said, Wayland is not a particularly complicated protocol. I don't know what kind of FFI Godot has, but I suppose that it _might_ be possible to implement some kind of plugin for Godot that uses wlroots or something to enable running a compositor in a game.

**Comment by ffernn-dev** [comment]
https://www.gamingonlinux.com/index.php?module=statistics&view=trends#SessionType-top
Looking at the statistics from this page (was linked to earlier but things have changed since), Wayland is on the way up pretty fast for gaming, and NVIDIA support is pretty good these days. I've been happily gaming and programming on Wayland for 6 months-ish now, with the only glaring issue at the moment being really awful rendering issues with apps running under Xwayland (Godot included, making it really hard to develop :disappointed: ).
At the risk of sounding like every linux user stereotype, I think Wayland is the way forward for linux desktops, as it's only getting better from here and I think most people would be happy to be rid of X11

**Comment by SopaDeMacaco-UmaDelicia** [comment]
Silly question. Why not just add SDL to Godot? It can help with running on Wayland natively. Many game engines (Unreal, Unity, Source, It Tech, use it, so it's defacto standard.

**Comment by ssokolow** [comment]
That's a shame. The #1 thing I value SDL for is something Godot fundamentally *cannot* replicate... the adoption by people like Valve as a means to inject updated platform support (Wayland, PipeWire, GNOME windeco, etc.) into statically linked games which have been abandoned by their developers via the `SDL_DYNAMIC_API` environment variable.

(That feature Ryan Gordon implemented where SDL's `Init` functions support performing runtime fixups to redirect the statically linked SDL's API surface to a newer, dynamic library copy of SDL.)

**Comment by SopaDeMacaco-UmaDelicia** [comment]
@ssokolow Yeah, that's a really nice feature. I was able to run old game (DEFCON) on wayland natively by just using sdl12-compat layer. SDL really did a great job with a backward compatibility.

**Comment by IntangibleMatter** [comment]
Considering there's talk of xorg being deprecated in the next few years, I think that having good Wayland support is becoming more and more urgent. Especially considering how many DEs are taking a wayland-first approach now or coming up.

**Comment by RushingAlien** [comment]
Yes that is true, notable plans are
- Fedora KDE to drop X11 session
- GNOME to drop x11 session support
- Budgie EFL rewrite will only have a Wayland session
- Cosmic DE by System76 will only have Wayland session

Some stuff related to gaming
- HDR work on Wayland is progressing

**Comment by Leshy-YA** [comment]
Wayland support is becoming more important as X11 is effectively a dead project. Godot is suffering from critical issues like https://github.com/godotengine/godot/issues/75308 because of this.

At this point it might be worth considering having Godot a not fully featured Wayland implementation available, so that X11 bugs, that nobody is likely to touch don't prevent people from working and having their projects running.

**Comment by Leshy-YA** [comment]
> @Leshy-YA we already have a WIP backend over at pr [godotengine/godot#57025](https://github.com/godotengine/godot/pull/57025). It's planned to be merged in 4.3, so it's only a matter of time, don't worry!

Reported: https://gitlab.freedesktop.org/xorg/lib/libx11/-/issues/199 I hope we can get a fix before Fedora 40.
